### ISSUE TITLE
slower than dense arrays

### ISSUE THREAD

**Issue by cscherrer** [JuliaArrays/ElasticArrays.jl#37]
I think I misunderstood what this package does. From the README I thought it might allocate memory in chunks to amortize the allocation cost. But 
```julia
julia> using BenchmarkTools

julia> t = Tuple((randn(), rand(1:1000), randn(), randn()))
(-0.43971567999268296, 972, -0.04411386377169432, 0.7367145291848813)

julia> dv = [t]
1-element Vector{Tuple{Float64, Int64, Float64, Float64}}:
 (-0.43971567999268296, 972, -0.04411386377169432, 0.7367145291848813)

julia> ev = ElasticVector([t])
1-element ElasticVector{Tuple{Float64, Int64, Float64, Float64}, 0, Vector{Tuple{Float64, Int64, Float64, Float64}}}:
 (-0.43971567999268296, 972, -0.04411386377169432, 0.7367145291848813)

julia> @btime push!($dv, $t);
  11.132 ns (0 allocations: 0 bytes)

julia> @btime push!($ev, $t);
  14.863 ns (0 allocations: 0 bytes)
```

Is it typical for this to be slower than a standard dense vector? Or maybe I'm doing something wrong?

**Comment by oschulz** [comment]
There's really no practical difference between a 1-dimensional `ElasticArray` and `Vector`, the performance is (and should be) basically the same.

```
julia> @btime for i in 1:1000 push!($dv, $t) end
  9.912 μs (0 allocations: 0 bytes)

julia> @btime for i in 1:1000 push!($ev, $t) end
  10.466 μs (0 allocations: 0 bytes)
```

What's special about `ElasticArray` is that a multi-dimensional one can grow in the last dimension (`Array{T,N}` can't be resized for `N > 1` at all). It doesn't use chunks though, everything is stored in a `Vector` internally.

**Comment by oschulz** [comment]
In combination with `ArraysOfArrays.VectorOfSimilarArrays`, ElasticArrays can be used to stored vectors of same-sized array in a contiguous fashion in memory (in one large block, not chunked).

**Comment by cscherrer** [comment]
Thanks @oschulz , this is very helpful. I agree it's especially useful with `ArraysOfArrays.VectorOfSimilarArrays`. I've been using it this way in SampleChains.jl:
```julia
julia> v = chainvec(randn(3))
1-element ArrayOfSimilarArrays{Float64, 1, 1, 2, ElasticMatrix{Float64, 1, Vector{Float64}}}:
 [0.03183382561830356, 0.2558030040983744, -0.8222026478748833]

julia> push!(v, randn(3))
2-element ArrayOfSimilarArrays{Float64, 1, 1, 2, ElasticMatrix{Float64, 1, Vector{Float64}}}:
 [0.03183382561830356, 0.2558030040983744, -0.8222026478748833]
 [-2.0387605821045596, 0.3338217450722492, 1.7606176401551503]
```

This is so useful, but it's not obvious until you've seen both packages, and also the types get kind of ugly. I've wondered about a type alias for this, or maybe a separate package for making this idiom more clear. Or maybe a `@require` in this package would make more sense?

Not sure, any thoughts on this?

[Off-topic from the OP, we can discuss elsewhere if you prefer]

**Comment by oschulz** [comment]
> I've been using it this way in SampleChains.jl

Fun fact, ElasticArrays.jl and ArraysOfArrays.jl were actually originally developed for exactly this use case: Compact storage of MCMC samples in BAT.jl. Then I forked it out as a separate package (but BAT.jl still uses them, of course).

> This is so useful, but it's not obvious until you've seen both packages [...] Or maybe a @require in this package would make more sense?

Since ElasticArrays.jl is pretty lightweight, I guess it could be added as a directly dependency to ArraysOfArrays.jl. I agree that it would be nice to have a convenience ctor for "resizable vector of arrays".

**Comment by oschulz** [comment]
Since we're OT anyhow: I had kinda missed your announcement of SampleChains.jl - very interesting, esp. since it seems that you support weighted samples. Do you also support a different number of samples per chain? Long-term, I could see BAT.jl using a standard packages like that (current it has it's own sample vector, https://github.com/bat/BAT.jl/blob/master/src/variates/density_sample.jl).

**Comment by oschulz** [comment]
Instead of tuples of field names, BAT.jl uses [ValueShapes.jl](https://github.com/oschulz/ValueShapes.jl), though (another package that was designed to play nice with ElasticArrays and ArraysOfArrays).

**Comment by cscherrer** [comment]
> Since we're OT anyhow: I had kinda missed your announcement of SampleChains.jl - very interesting, esp. since it seems that you support weighted samples.

Weighted samples are in very early stages, but yes I think it's an important case to support. Currently the farthest along is [SampleChainsDynamicHMC](https://github.com/cscherrer/SampleChainsDynamicHMC.jl). @mschauer and I are currently working out how best to represent continuous trajectories for his [ZigZagBoomerang](https://github.com/mschauer/ZigZagBoomerang.jl) package. He has the core of this figured out, just a matter of making sure the abstraction and the interface is consistent across all SampleChains instances.

I hadn't seen BAT.jl, I'll need to look into it! In general I'd be very open to collaboration on this stuff, especially wrt SampleChains, MeasureTheory, and Soss, and any of the ecosystem affecting those.

> Do you also support a different number of samples per chain?

Yes, a `MultiChain` just holds a collection of `SampleChain`s, there's no constraint for them to have the same length. I don't have anything going in parallel yet, but I think it's natural to want the chains to stop at the same time, rather than after a fixed number of samples.

> Instead of tuples of field names, BAT.jl uses ValueShapes.jl, though (another package that was designed to play nice with ElasticArrays and ArraysOfArrays).

Looks like I'll need to dig into that as well, looks like it maybe we could come to something with the best of that and the [TupleVectors](https://github.com/cscherrer/SampleChains.jl/blob/main/src/tuplevectors.jl) I built for SampleChains.jl.

**Comment by oschulz** [comment]
> I hadn't seen BAT.jl

It's been around for a while, I just always think "let's add this one thing then I'll write an announcement" :-)

> Looks like I'll need to dig into that as well

ValueShapes also support array-shapes parameters and parameters pinned to constant values. Also it includes a `NamedTupleDist` that we use as the main tool to specify priors in BAT.jl. It implements actual `NamedTuple`-valued distributions that can be flattened the same way as the sample variates.

**Comment by oschulz** [comment]
If you like, take a look here: https://bat.github.io/BAT.jl/stable/tutorial/#Prior-Definition

I'm very open to collaboration here as well.

**Comment by cscherrer** [comment]
Looks like there's a lot of commonality here. Want to chat some more in Zulip?

**Comment by oschulz** [comment]
Sure!
